### ISSUE TITLE
BlockingHttp4sContext doesn't properly handle shifted IO

### ISSUE THREAD

**Issue by nigredo-tori** [http4s/http4s#2939]
I've tried using `BlockingHttp4sContext` with Jetty, and found that my response bodies were not sent, with things like this in the log:
<details><pre>
16:09:06.585 [qtp1582668443-7357] WARN  org.eclipse.jetty.io.FillInterest - Read pending for org.eclipse.jetty.server.HttpConnection$BlockingReadCallback@c7e5277 prevented AC.ReadCB@41fe7925{HttpConnection@41fe7925::SocketChannelEndPoint@36cf27e4{/127.0.0.1:35648<->/127.0.0.1:42993,OPEN,fill=FI,flush=-,to=3/30000}{io=1/1,kio=1,kro=1}->HttpConnection@41fe7925[p=HttpParser{s=START,0 of -1},g=HttpGenerator@41d4b1b5{s=START}]=>HttpChannelOverHttp@6018ab31{r=1,c=false,c=false/false,a=IDLE,uri=null,age=0}}
16:09:06.595 [qtp1582668443-7357] WARN  o.e.j.util.thread.QueuedThreadPool - 
java.nio.channels.ReadPendingException: null
	at org.eclipse.jetty.io.FillInterest.register(FillInterest.java:58)
	at org.eclipse.jetty.io.AbstractEndPoint.fillInterested(AbstractEndPoint.java:362)
	at org.eclipse.jetty.io.AbstractConnection.fillInterested(AbstractConnection.java:134)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:276)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:782)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:918)
	at java.base/java.lang.Thread.run(Thread.java:834)
16:09:06.666 [http4s-spec-3] ERROR org.http4s.servlet.Http4sServlet - Error processing request
java.nio.channels.ClosedChannelException: null
	at org.eclipse.jetty.io.FillInterest.onClose(FillInterest.java:150)
	at org.eclipse.jetty.io.AbstractEndPoint.onClose(AbstractEndPoint.java:355)
	at org.eclipse.jetty.io.ChannelEndPoint.onClose(ChannelEndPoint.java:214)
	at org.eclipse.jetty.io.AbstractEndPoint.doOnClose(AbstractEndPoint.java:225)
	at org.eclipse.jetty.io.AbstractEndPoint.close(AbstractEndPoint.java:192)
	at org.eclipse.jetty.io.AbstractEndPoint.close(AbstractEndPoint.java:175)
	at org.eclipse.jetty.io.AbstractConnection.close(AbstractConnection.java:254)
	at org.eclipse.jetty.io.ManagedSelector.closeNoExceptions(ManagedSelector.java:254)
	at org.eclipse.jetty.io.ManagedSelector.access$1400(ManagedSelector.java:61)
	at org.eclipse.jetty.io.ManagedSelector$CloseConnections.update(ManagedSelector.java:867)
	at org.eclipse.jetty.io.ManagedSelector$SelectorProducer.processUpdates(ManagedSelector.java:428)
	at org.eclipse.jetty.io.ManagedSelector$SelectorProducer.produce(ManagedSelector.java:399)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produceTask(EatWhatYouKill.java:360)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:184)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produce(EatWhatYouKill.java:135)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:782)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:918)
	at java.base/java.lang.Thread.run(Thread.java:834)
	Suppressed: java.lang.Throwable: HttpInput failure
		at org.eclipse.jetty.server.HttpInput.failed(HttpInput.java:822)
		at org.eclipse.jetty.server.HttpConnection$BlockingReadCallback.failed(HttpConnection.java:659)
		... 19 common frames omitted
</pre></details>
or
<details><pre>
16:40:51.659 [http4s-spec-0] ERROR org.http4s.servlet.Http4sServlet - Error processing request after response was committed
java.io.IOException: org.eclipse.jetty.http.BadMessageException: 500: No version
	at org.eclipse.jetty.server.HttpOutput.write(HttpOutput.java:236)
	at org.eclipse.jetty.server.HttpOutput.write(HttpOutput.java:551)
	at org.http4s.servlet.BlockingServletIo$$anonfun$initWriter$1$$anonfun$apply$1.apply(ServletIo.scala:52)
	at org.http4s.servlet.BlockingServletIo$$anonfun$initWriter$1$$anonfun$apply$1.apply(ServletIo.scala:49)
	at fs2.Chunk$Singleton.map(Chunk.scala:489)
	at fs2.internal.Algebra$$anon$1.apply(Algebra.scala:68)
	at fs2.internal.Algebra$$anon$1.apply(Algebra.scala:66)
	at fs2.internal.FreeC$Eval$$anonfun$translate$2.apply(FreeC.scala:179)
	at fs2.internal.FreeC$Eval$$anonfun$translate$2.apply(FreeC.scala:179)
	at fs2.internal.FreeC$$anonfun$suspend$1.apply(FreeC.scala:193)
	at fs2.internal.FreeC$$anonfun$suspend$1.apply(FreeC.scala:193)
	at fs2.internal.FreeC$$anonfun$flatMap$1.apply(FreeC.scala:33)
	at fs2.internal.FreeC$$anonfun$flatMap$1.apply(FreeC.scala:30)
	at fs2.internal.FreeC$ViewL$$anonfun$mk$1.apply(FreeC.scala:216)
	at fs2.internal.FreeC$ViewL$$anonfun$mk$1.apply(FreeC.scala:216)
	at fs2.internal.FreeC$ViewL$.mk(FreeC.scala:214)
	at fs2.internal.FreeC$ViewL$.apply(FreeC.scala:206)
	at fs2.internal.FreeC.viewL(FreeC.scala:67)
	at fs2.internal.Algebra$.fs2$internal$Algebra$$go$1(Algebra.scala:214)
	at fs2.internal.Algebra$$anonfun$fs2$internal$Algebra$$go$1$4$$anonfun$apply$5$$anonfun$apply$7.apply(Algebra.scala:260)
	at fs2.internal.Algebra$$anonfun$fs2$internal$Algebra$$interruptGuard$1$1.apply(Algebra.scala:230)
	at fs2.internal.Algebra$$anonfun$fs2$internal$Algebra$$interruptGuard$1$1.apply(Algebra.scala:229)
	at cats.effect.internals.IORunLoop$.cats$effect$internals$IORunLoop$$loop(IORunLoop.scala:139)
	at cats.effect.internals.IORunLoop$.startCancelable(IORunLoop.scala:41)
	at cats.effect.internals.IOBracket$BracketStart.run(IOBracket.scala:86)
	at cats.effect.internals.Trampoline.cats$effect$internals$Trampoline$$immediateLoop(Trampoline.scala:70)
	at cats.effect.internals.Trampoline.startLoop(Trampoline.scala:36)
	at cats.effect.internals.TrampolineEC$JVMTrampoline.cats$effect$internals$TrampolineEC$JVMTrampoline$$super$startLoop(TrampolineEC.scala:93)
	at cats.effect.internals.TrampolineEC$JVMTrampoline$$anonfun$startLoop$1.apply$mcV$sp(TrampolineEC.scala:93)
	at cats.effect.internals.TrampolineEC$JVMTrampoline$$anonfun$startLoop$1.apply(TrampolineEC.scala:93)
	at cats.effect.internals.TrampolineEC$JVMTrampoline$$anonfun$startLoop$1.apply(TrampolineEC.scala:93)
	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:72)
	at cats.effect.internals.TrampolineEC$JVMTrampoline.startLoop(TrampolineEC.scala:92)
	at cats.effect.internals.Trampoline.execute(Trampoline.scala:43)
	at cats.effect.internals.TrampolineEC.execute(TrampolineEC.scala:44)
	at cats.effect.internals.IOBracket$BracketStart.apply(IOBracket.scala:72)
	at cats.effect.internals.IOBracket$BracketStart.apply(IOBracket.scala:52)
	at cats.effect.internals.IORunLoop$.cats$effect$internals$IORunLoop$$loop(IORunLoop.scala:136)
	at cats.effect.internals.IORunLoop$.start(IORunLoop.scala:34)
	at cats.effect.internals.IOBracket$$anonfun$apply$1.apply(IOBracket.scala:44)
	at cats.effect.internals.IOBracket$$anonfun$apply$1.apply(IOBracket.scala:34)
	at cats.effect.internals.IORunLoop$RestartCallback.start(IORunLoop.scala:341)
	at cats.effect.internals.IORunLoop$.cats$effect$internals$IORunLoop$$loop(IORunLoop.scala:119)
	at cats.effect.internals.IORunLoop$RestartCallback.signal(IORunLoop.scala:355)
	at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:376)
	at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:316)
	at cats.effect.internals.IOShift$Tick.run(IOShift.scala:36)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: org.eclipse.jetty.http.BadMessageException: 500: No version
	at org.eclipse.jetty.http.HttpGenerator.generateResponse(HttpGenerator.java:389)
	at org.eclipse.jetty.server.HttpConnection$SendCallback.process(HttpConnection.java:743)
	at org.eclipse.jetty.util.IteratingCallback.processing(IteratingCallback.java:241)
	at org.eclipse.jetty.util.IteratingCallback.iterate(IteratingCallback.java:223)
	at org.eclipse.jetty.server.HttpConnection.send(HttpConnection.java:549)
	at org.eclipse.jetty.server.HttpChannel.sendResponse(HttpChannel.java:852)
	at org.eclipse.jetty.server.HttpChannel.write(HttpChannel.java:929)
	at org.eclipse.jetty.server.HttpOutput.write(HttpOutput.java:250)
	at org.eclipse.jetty.server.HttpOutput.write(HttpOutput.java:226)
	... 49 common frames omitted
</pre></details>

Basically, in some cases `BlockingHttp4sServlet` handles the response asynchronously. Since `HttpServlet.service` method returned before the request was properly handled, Jetty sent the 200 status with an empty body.

I've made a small fix for the 0.20 branch, and will send a PR shortly. It should be easy to port that to 0.21 later, but 0.20 is a priority for me right now (I have to fix this for a Scala 2.11 project).

BTW, I had a similar (?) issue in production with the asynchoronous servlet (servlet writing to a committed response), and making a fix for that looked much harder. I can't reproduce that failure locally, so I have to focus on the blocking servlet for now.

**Comment by hamnis** [comment]
@nigredo-tori this has been fixed? Close?
